### PR TITLE
[KEYCLOAK-16373] Document the streamified variant of the user storage…

### PR DESCRIPTION
**Pull request by sguilhen**
… interfaces

### CODE REVIEW

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+To make use of the `Streams` variants, one has to simply implement these sub-interfaces instead of the original interfaces
+when creating a custom provider. For example, the following code uses the `Streams` variant of the `UserQueryProvider` interface:
+
+[source,java]
+----
+public class CustomQueryProvider extends UserQueryProvider.Streams {
+...
+    @Override
+    Stream<UserModel> getUsersStream(RealmModel realm, int firstResult, int maxResults) {
```
Comment (martin-kanis) [inline server_development/topics/user-storage/stream-interfaces.adoc]:
```suggestion
    Stream<UserModel> getUsersStream(RealmModel realm, Integer firstResult, Integer maxResults) {
```

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+|===
+| Package | Classes
+| `org.keycloak.credential` | `CredentialInputUpdater`(*), `UserCredentialStore`
+| `org.keycloak.models` | `GroupModel`, `RoleMapperModel`, `UserCredentialManager`, `UserModel`, `UserProvider`
+| `org.keycloak.models.cache` | `CachedUserModel`, `UserCache`
+| `org.keycloak.storage.federated` | All classes
+| `org.kecyloak.storage.user` | `UserQueryProvider`(*)
+|===
+(*) identify _capability interfaces_
```
Comment (martin-kanis) [inline server_development/topics/user-storage/stream-interfaces.adoc]: I don't understand what you mean by this note :)

**Review by martin-kanis — CHANGES_REQUESTED** [review]
Thank you Stefan for the PR. Overall it looks good, I just put two small comments inline.
I would like to let @andymunro to review this from language perspective. 

I noticed there are again failures regarding wrong links and missing picture. I don't see how it's related to your changes. I thought the problems were fixed.

File: server_development/topics/user-storage/stream-interfaces.adoc (hunk 1 above)
Comment (sguilhen) [reply server_development/topics/user-storage/stream-interfaces.adoc]: Ah, yes.. I was using my last worked-on branch which doesn't have these changes.. I'm updating the example

File: server_development/topics/user-storage/stream-interfaces.adoc (hunk 2 above)
Comment (sguilhen) [reply server_development/topics/user-storage/stream-interfaces.adoc]: At the beginning of the User Storage SPI chapter there's a section called "Provider Capability Interfaces", which are basically interfaces a custom storage would implement if it needs the capabilities (or features) provided by those particular interfaces. In this list I wanted to highlight that a couple of the interfaces that offer the Streams variant are capability interfaces, which are more likely to be implemented by a custom provider when compared to other interfaces.

**Comment by sguilhen** [comment]
Thanks for the review, Martin!

> Thank you Stefan for the PR. Overall it looks good, I just put two small comments inline.
> I would like to let @andymunro to review this from language perspective.
> 
> I noticed there are again failures regarding wrong links and missing picture. I don't see how it's related to your changes. I thought the problems were fixed.

Yeah, not triggered by my changes. I've sent another PR (https://github.com/keycloak/keycloak-documentation/pull/1087) to fix the broken links (images are a different thing that needs to be checked with Andy)

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 3]
```diff
@@ -0,0 +1,48 @@
+
+=== Stream-based Interfaces
+
+Many of the user storage interfaces in {project_name} contain query methods that can return potentially large sets of objects,
+which might lead to in significant impacts in terms of memory consumption and processing time - specially when only a small subset of the
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: delete "in"
change "time - specially" to "time.  This is especially true..."

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+objects' internal state is used in the query method's logic.
+
+In order to provide developers with a more efficient alternative to process large data sets in these query methods, a `Streams`
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: Delete "In order"

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+sub-interface has been added to user storage interfaces. These `Streams` sub-interfaces replace the original collection-based
+methods in the super-interfaces with stream-based variants, making the collection-based methods default (they invoke their `Stream`
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: Make parenthetical comment a separate sentence.

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+counterpart and collect the result).
+
+The `Streams` sub-interfaces allows for implementations to focus on the stream-based approach for processing sets of data and
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: allow not allows

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+benefit from the potential memory and performance optimizations of that approach. The interfaces that offer a `Streams`
+sub-interface to be implemented include a few _capability interfaces_, all interfaces in the `org.keycloak.storage.federated`
+package and a few others that might be implemented depending on the scope of the custom storage implementation.
+
+You can find bellow a list of the interfaces that offer a `Streams` sub-interface to developers.
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: Change "
You can find bellow a"
to 
"See this"

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+| Package | Classes
+| `org.keycloak.credential` | `CredentialInputUpdater`(*), `UserCredentialStore`
+| `org.keycloak.models` | `GroupModel`, `RoleMapperModel`, `UserCredentialManager`, `UserModel`, `UserProvider`
+| `org.keycloak.models.cache` | `CachedUserModel`, `UserCache`
+| `org.keycloak.storage.federated` | All classes
+| `org.kecyloak.storage.user` | `UserQueryProvider`(*)
+|===
+(*) indicates the interface is a _capability interface_
+
+To make use of the `Streams` variants, one has to simply implement these sub-interfaces instead of the original interfaces
```
Comment (andymunro) [inline server_development/topics/user-storage/stream-interfaces.adoc]: change "one has to simply implement" to "use"

**Review by andymunro — CHANGES_REQUESTED** [review]
Thanks, Stefan.  A few changes.

**Comment by sguilhen** [comment]
Thanks @andymunro! Just pushed the docs again after incorporating your suggestions.

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+|===
+| Package | Classes
+| `org.keycloak.credential` | `CredentialInputUpdater`(*), `UserCredentialStore`
+| `org.keycloak.models` | `GroupModel`, `RoleMapperModel`, `UserCredentialManager`, `UserModel`, `UserProvider`
+| `org.keycloak.models.cache` | `CachedUserModel`, `UserCache`
+| `org.keycloak.storage.federated` | All classes
+| `org.kecyloak.storage.user` | `UserQueryProvider`(*)
+|===
+(*) indicates the interface is a _capability interface_
```
Comment (hmlnarik) [inline server_development/topics/user-storage/stream-interfaces.adoc]: Can the `_capability interface_` be a link to `Provider Capability Interfaces` section?

File: server_development/topics/user-storage/stream-interfaces.adoc [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+See this list of the interfaces that offer a `Streams` sub-interface to developers.
+
+|===
+| Package | Classes
+| `org.keycloak.credential` | `CredentialInputUpdater`(*), `UserCredentialStore`
+| `org.keycloak.models` | `GroupModel`, `RoleMapperModel`, `UserCredentialManager`, `UserModel`, `UserProvider`
+| `org.keycloak.models.cache` | `CachedUserModel`, `UserCache`
+| `org.keycloak.storage.federated` | All classes
+| `org.kecyloak.storage.user` | `UserQueryProvider`(*)
+|===
```
Comment (hmlnarik) [inline server_development/topics/user-storage/stream-interfaces.adoc]: This list needs to respect community vs product build limitations, e.g. `CachedUserModel` should not appear in the product documentation.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
@sguilhen Thanks for the contribution! Please see inline

File: server_development/topics/user-storage/stream-interfaces.adoc (hunk 9 above)
Comment (sguilhen) [reply server_development/topics/user-storage/stream-interfaces.adoc]: Done

File: server_development/topics/user-storage/stream-interfaces.adoc (hunk 10 above)
Comment (sguilhen) [reply server_development/topics/user-storage/stream-interfaces.adoc]: Checked the product interface. I've removed the cache.* interfaces from product, along with UserCredentialStore, UserCredentialManager and UserProvider. All other interfaces are mentioned as implementable earlier in the prod docs.

**Review by hmlnarik — APPROVED** [review]

**Review by martin-kanis — APPROVED** [review]

**Pull request merged**